opal/mca/btl/ofi/btl_ofi_component.c

```diff
@@ -347,11 +353,28 @@ static mca_btl_base_module_t **mca_btl_ofi_component_init(int *num_btl_modules,
     /* Request device transfer capabilities, separate from required_caps */
     hints.caps |= FI_HMEM;
     hints.domain_attr->mr_mode |= FI_MR_HMEM;
+
+    /*
+     * Libfabric providers choose whether to perform CUDA memory transfers when
+     * FI_HMEM_CUDA_ENABLE_XFER is unspecified. Some providers may choose to
+     * disable CUDA memory transfers by default. Explicitly enabling this option
+     * allows providers to perform CUDA memory transfers via FI_HMEM if
+     * supported.
+     */
+    if ((false != opal_common_ofi.hmem_cuda_enable_xfer) &&
+        (OPAL_EXISTS != opal_setenv("FI_HMEM_CUDA_ENABLE_XFER", "1", false, &environ))) {
+        unset_fi_hmem_cuda_enable_xfer = true;
+    }
```

Comment:
Per the discussion on this PR about horror stories from users not getting expected behavior, can we at least ensure that libfabric is obeying the setting of this MCA parameter?  I can think of two cases offhand (there may be more?):

1. OMPI should emit a warning (or possibly even error?) if `opal_common_ofi.hmem_cuda_enable_xfer` is true but `FI_HMEM_CUDA_ENABLE_XFER` is already set in the environment to a "false" value.  I.e., the OMPI user asked for XYZ, but the environment is set for ABC (and therefore OMPI won't overwrite it).  OMPI absolutely should not silently ignore this situation -- it should tell the user, "You asked for XYZ, but I couldn't do that."  OMPI has many precedents for "A user asked me to do XYZ, but I could not do that, so abort a let a human figure it out."

2. Since the EFA provider may have already set the CUDA HMEM behavior (e.g., in a prior call from elsewhere in the application), I think your text elsewhere in this PR indicates that libfabric may ignore the value of the `FI_HMEM_CUDA_ENABLE_XFER` env variable in *this* call to `fi_getinfo()`.  Is there a way to tell?  I.e., in the output that you get here from `fi_getinfo()`, can you tell if `FI_HMEM_CUDA_ENABLE_XFER` was honored or not?  Again, if an OMPI user asks for behavior XYZ but Open MPI is not able to get that behavior, it should -- at a bare minimum -- emit a warning about that.

Reply:
It might also be a good idea to find a way to limit this to the EFA provider, or vendors that opt "in". Seems a shame to create the customer relation problems for everyone if they haven't bought into it. Perhaps an MCA param that controls whether or not to enable this behavior? Just trying to find a way to limit the damage to those vendors who accept the risk.

Reply:
> To put it plainly, FI_HMEM_CUDA_ENABLE_XFER=1 is intended specifically for CUDA-aware MPI workloads.

It is my impression that CUDA-aware MPI is pretty common nowadays. My point was that you are setting this behavior for everyone, even if not everyone would agree with your assertion:
```c
    * Libfabric providers choose whether to perform CUDA memory transfers when
     * FI_HMEM_CUDA_ENABLE_XFER is unspecified. Some providers may choose to
     * disable CUDA memory transfers by default. Explicitly enabling this option
     * allows providers to perform CUDA memory transfers via FI_HMEM if
     * supported.
```

> We can squint and see this as a flaw in user experience (which is fair), but I believe the blast radius of this change is being overstated a bit.

From personal experience, I think you are underestimating the reaction. What I'm suggesting is that you make this an "opt in" selection - i.e., only those who understand and accept the risks should be impacted by this decision.

You could perhaps set the envar in your Amazon images - you don't need to set it in OMPI itself as you control the images. This would achieve the same affect as this PR, but limit it to just Amazon. Other organizations who want to do the same thing could follow a similar route using the default MCA param file, which allows the setting of default envars.

Guess my bottom line is: why hardwire this in code when we have default envars you can set in your image or in the default MCA param file, both of which can be overridden by the user?

Reply:
> From personal experience, I think you are underestimating the reaction. What I'm suggesting is that you make this an "opt in" selection - i.e., only those who understand and accept the risks should be impacted by this decision.

I don't want to downplay the risks at hand, or the potential precedent being set by utilizing an env var to control downstream implementation details (in this case: is `cudaMemcpy()` safe to use?). In Libfabric, `FI_HMEM_CUDA_ENABLE_XFER` was `true` by default prior to ofiwg/libfabric#8354 and we'd like to maintain this behavior going forward for Open MPI users. Making this "opt in" only defeats the purpose of this PR, which is to absolve the user from needing to add `FI_HMEM_CUDA_ENABLE_XFER=1`.

> You could perhaps set the envar in your Amazon images - you don't need to set it in OMPI itself as you control the images. This would achieve the same affect as this PR, but limit it to just Amazon

This would present the conflict we're ultimately trying to avoid: we want to prevent Libfabric from performing CUDA memory transfers for NCCL users on AWS (who represent the majority of our business); setting the env var in the image itself would not be feasible here. We also obviously want to support OMPI users on non-Amazon-managed images.

Reply:
So why not just put the envar in the default MCA param file for NCCL users? Or tell NCCL users that they need to set the envar? Or since (IIRC) NCCL involves something being put in the environment, perhaps detect a NCCL envar and invoke the behavior in only that case?

Just trying to find a more targeted mechanism here so you don't hardcode something into OMPI that really only affects your own community, or affects only one subset of the overall OMPI user-base. We have created several mechanisms specifically to deal with such targeted issues - surely one of those (or perhaps a minor revision to one of those) can be employed here instead of a hardwired method that hits everywhere?

Overall, my concern is motivated by the seeming acceptance of a faulty mechanism for controlling this behavior, and then further injecting that problem into OMPI. If we keep the problem at a level above the OMPI code, then we (a) deflect the blame when users are impacted by it back onto the libfabric community, which ultimately is responsible for the error; and (b) retain flexibility in OMPI to sidestep the problem as users encounter it.

If we can find an external method for dealing with it, then we avoid pulling OMPI into the middle of the mess. Advice to users is one way of doing it - simply point out the flaw in libfabric, advise them of the potential pitfalls, and suggest they at least try to deal with it by setting the envar prior to invoking their application. This at least provides an attempt at mitigating the damage and leaves you in the best light. You can even note that you are working with the libfabric community to resolve the issue (which you really should be doing).

Since some (most?) users won't read the advice, then you try setting default envars to help them. Probably your best bet is to just leave libfabric to its default setting, perhaps printing out a warning if you detect that it could have been set differently once you know the underlying hardware and its capabilities. I suspect a couple of times is all that would be required before the user started setting params the way they want it to behave. Still protects OMPI (and you) from someone unexpectedly getting odd behavior due to envar overwrites.

Again, I'm probing for a less intrusive solution than what is proposed here. I'd ultimately like to see something that doesn't expose OMPI to customer anger, even if it means emitting warnings and advice to users (along with a param to shut it off).

Reply:
> Or tell NCCL users that they need to set the envar?

This path has more resistance than simply setting the env var in OMPI ahead of a 5.0.x release to maintain the old behavior.

> Overall, my concern is motivated by the seeming acceptance of a faulty mechanism for controlling this behavior, and then further injecting that problem into OMPI. If we keep the problem at a level above the OMPI code, then we (a) deflect the blame when users are impacted by it back onto the libfabric community, which ultimately is responsible for the error; and (b) retain flexibility in OMPI to sidestep the problem as users encounter it.

Would clear warning messages suffice? As Jeff proposed, warning or failing when either the MCA param or env var conflict? I fully empathize with the burden of fielding potential complaints from users for Libfabric issues.

> You can even note that you are working with the libfabric community to resolve the issue (which you really should be doing)

There's nothing really to resolve on our end. As mentioned elsewhere, we're using the env var `FI_HMEM_CUDA_ENABLE_XFER` because it's pre-existing and would ostensibly conflict with any new mechanism. That's the gist of it. Otherwise we would have chosen something else.

> envar overwrites

To be pedantic again, the logic in this PR does not overwrite the var. The env var is only set (and subsequently unset) if it doesn't exist already. The initial commit I pushed did overwrite it, but I changed that after Jeff's feedback.

Reply:
> You can even note that you are working with the libfabric community to resolve the issue (which you really should be doing)

> There's nothing really to resolve on our end.

Perhaps that is what is bothering me the most - it is the acceptance that a flawed mechanism should not be confronted and fixed. What you are proposing here is a bandaid that only pushes the problem down the road. If we are going to accept a bandaid, then it would be nice to have some assurance that you are addressing the core problem.

Otherwise, this is a perpetual problem that puts OMPI in the middle of it.

> This path has more resistance than simply setting the env var in OMPI ahead of a 5.0.x release to maintain the old behavior.

I get that you are overriding the new libfabric default to restore prior behavior - but that in itself is controversial for those expecting the new default libfabric behavior.(since OMPI hasn't released v5 yet). I get that Amazon thinks that new default isn't right for its customers, but that might not match what others believe, nor might it be a permanent situation at Amazon (e.g., if NCCL or libfabric changes at some point - this code will be out there for quite some time, so we are stuck with whatever gets hardcoded here).

Ultimately, you are placing your customers ahead of the rest of the community - i.e., rather than inconvenience them, you inconvenience everyone else who has a different take on what the libfabric setting should be. Understood from your perspective, but I think you can appreciate the pushback.

> Would clear warning messages suffice? As Jeff proposed, warning or failing when either the MCA param or env var conflict? I fully empathize with the burden of fielding potential complaints from users for Libfabric issues.

Since you don't know what someone might previously have set and then unset (just as you are doing here), it seems like another frail point-of-failure. Bottom line is you cannot know that you are actually setting anything, nor can you know if the envar previously existed, or will exist at some later point. You're just taking a shot in the dark and hoping to get lucky.

I'm not sure how any warning message can capture those cases - sounds awfully fragile to me.

Reply:
@rhc54

I think there are some miscommunication here, and let me clarify:

Ultimately, the reason for Open MPI to set this environment behavior is because Open MPI follow this doc:
         https://docs.nvidia.com/cuda/gpudirect-rdma/#sync-behavior
to ensure memory consistency regarding GDR.

Note that not all customers of libfabric is following this guidance (NCCL, for example, does not follow this doc, for the sake of efficiency). 

Libfabric need to know whether a customer is following this guidance. Therefore libfabric provided this environment variable. Setting it to true means the user is following this guidance.

Because open mpi chose to follow this guidance, it is natural to think that openmpi set this environment variable, would you agree?

A user of openmpi does not know the detail of how openmpi is implementing GDR transfer, so it would not be user's responsibility to set this environment variable.

Reply:
The problem is that you have no way of knowing if you actually set that variable inside libfabric. If it was set ahead of you, then your operation is ignored. Likewise, there is no way for a user or application (or another library) to know whether or not you set it.

So the issue here is that you set an envar and then init the library - and make the completely unverifiable assumption that you have set libabric to the behavior you claim to be following. This is ultimately the objection. OMPI is now making an indeterministic assumption about libfabric behavior, and even worse, is claiming to its users that this _is_ the behavior. If the OMPI internal code is actually relying on that setting, then we are in significant trouble.

What I'm advocating you do is (a) correctly communicate to your users that this is the situation with respect to libfabric usage, (b) don't hardwire this code in OMPI, but instead use our already well-built methods for setting this envar that deal with overwrite conditions, and (c) take the indeterminism problem back to the libfabric community to be resolved. Yes, I realize this inconveniences your customers somewhat, and it involves more work on the libfabric elements of the OMPI community...but I don't see how to avoid it.

Otherwise, this is a crapshoot - if you are lucky, you win, but you are just as likely to lose without knowing it. Meantime, you push that indeterminism into the entire OMPI user base - which I suspect will not be appreciated.

Reply:
> The problem is that you have no way of knowing if you actually set that variable inside libfabric. If it was set ahead of you, then your operation is ignored. Likewise, there is no way for a user or application (or another library) to know whether or not you set it.

Sorry, I am confused about who are "you" in this paragraph?

Reply:
If I'm understanding correctly, you'd prefer to see Libfabric provide a clean mechanism to enable CUDA memory transfers (perhaps a capability bit for `hints->caps` while officially deprecating `FI_HMEM_CUDA_ENABLE_XFER`) and deal with any conflicting logic in Libfabric's internal initialization (e.g. `FI_HMEM_CUDA_ENABLE_XFER` being set in the env, but the not in `hints->caps`, or vice-versa). Then this PR would simply be something like:

```diff
     hints->caps = FI_HMEM;
+    hints->caps = FI_HMEM_CUDA_XFER;
```

...and Open MPI doesn't need to concern itself at all with setting and unsetting an environment variable based on various conditions - which aren't even comprehensive enough to cover all the different scenarios an env var could be set/unset. Does that sound right or am I off-base?

Reply:
Correct - I think that would resolve the determinism problem and avoid the future issues. Probably also need a query capability so you can determine if/how it was already set, and if someone set it. The basic concepts of your proposed change (the tri-state logic, for example) aren't bad - it's the indeterminism that is the root problem, and the fact that it hits all providers and users regardless of their circumstances (e.g., NCCL or not, CUDA-aware MPI or not, etc.).

Reply:
> Sorry, I am confused about who are "you" in this paragraph?
> 
> libfabric does not set the environment variable, it read the environment variable.
> 
> Customer of llibfabric (like mtl/ofi) set the environment variable, and it know it set the envvar.

I'm sorry - the "you" in this case is OMPI. This PR attempts to set an envar and uses that to direct an internal libfabric setting. My point is that you (OMPI) have no idea if your attempt was successful - if some other external actor (the app, the user, or another library that uses libfabric) previously set the libfabric behavior, then they win and you have no way of knowing that situation exists. Same problem exists for those other actors - they cannot know that OMPI has done this, and can only wonder with frustration when libfabric doesn't behave as they are telling it (unsuccessfully) to do.

This indeterminism is the root cause of the problems the libfabric mechanism exposes - and was the root cause of the problems in prior similar approaches in other libraries. It leads to the bad behavior and customer experiences detailed above. Hence the desire to at least minimize the exposure, but also to ensure that we (the OMPI community) use our leverage in the libfabric community to fix the underlying root cause problem.

Otherwise, we'll be dealing with the consequences for quite some time.

Reply:
To summarize (for my own sake):
- The response information returned by Libfabric (`struct fi_info **info` passed to `fi_getinfo()`) should provide some kind of feedback wrt whether or not the CUDA memory transfers were successfully enabled (or disabled). This is independent of the actual interface (whether it's an env var, capability bit, etc). Without this, OMPI has little (or no) confidence that everything will work as expected.
- Ideally, Libfabric would utilize something other than an env var as the control mechanism/interface. Even if all agree that an Open MPI user has no use for `FI_HMEM_CUDA_ENABLE_XFER=0`, there is no truly foolproof way to set `FI_HMEM_CUDA_ENABLE_XFER=1` without potentially triggering something unintended (no matter how unlikely).
    - If you set `FI_HMEM_CUDA_ENABLE_XFER=1` unconditionally (`overwrite = true`), you potentially mutate the user's environment behind their back.
    - If you set `FI_HMEM_CUDA_ENABLE_XFER=1` only if it doesn't exist in the environment, this doesn't take into account some other layer in the stack hypothetically setting `FI_HMEM_CUDA_ENABLE_XFER=0`. This would prevent CUDA memory transfers (inadvertently or otherwise), which is obviously poor user experience and likely kicks off a fun debugging session for all involved parties, with Open MPI naturally being the prime suspect. I would even argue the more logging/reporting Open MPI outputs wrt `FI_HMEM_CUDA_ENABLE_XFER`, the more likely they'd be to blame for any issues related to it.

Ultimately, we're asking Open MPI to put itself in a bit of an unfair position due to a design quirk/flaw in Libfabric. The more I think about it, the more it feels like lipstick on a pig; no amount of robustness in logic (Checking the environment, MCA params, restricting to certain providers, etc etc) can ever really give OMPI full confidence that everything will work as intended. Thus OMPI can't report to users with full confidence that everything will work as intended.

Software bugs and design flaws bubble up over time long after features are intially implemented, and you always end up wishing things were done right the first time. So I do understand the pushback when flaws stare you in the face and the fallout feels inevitable.

Reply:
Yeah, I feel bad that you (and really all of the OMPI community) has been put in this position. I deeply appreciate your patience with all of this. None of the pushback is due to your work - as I said, the ideas put forward in this PR are actually rather good. You just don't have a reliable mechanism by which to execute them.

Some of us may be able to help explain this to the libfabric folks and help them to understand the dilemma we are in. Probably best starts with someone filing an issue over there to begin the discussion. I have already reached out to another organization that went thru the prior similar problem I mentioned (fwiw, that was also an Intel library) and also uses libfabric, so they may step in to help as well. I'll continue to rally help from outside OMPI.

Reply:
@darrylabbate That was an excellent summary; thank you.  And thanks for your patience!

Also, thanks for clarifying that the issue is not EFA-specific -- that `FI_HMEM_CUDA_ENABLE_XFER` is a libfabric-wide mechanism.  Sorry if I mis-characterized it.

One other point/clarification: I think we've mostly been discussing the non-determinism of the mechanism to enable/disable this behavior.  I do *not* think that we have been discussing what Open MPI's default behavior should be (i.e., whether to enable or disable the libfabric HMHM cuda xfer stuff).  That is a different discussion, and should be had among those who care about OFI support in Open MPI (Cisco/usnic doesn't support this libfabric mechanism, so we don't care).  It's no different than any other network-stack-specific feature: those who care should discuss and decide what the default behavior should be.

Reply:
@jsquyres

I wonder if there is a guidance regarding the criteria to call `setenv` from inside Open MPI.

If I understand correctly, the concern @rhc54 raised regarding non-determinism of setting environment variable applies to any library that try to use envar to customize its behavior, not just libfabric.

When I look through the Open MPI code base, I can see current Open MPI code  call `setenv` (or `opal_setenv`) to set similar envars for PSM, PSM2, HCOLL, PVFS etc.

It is unclear to me whether or the same non-determinism concern apply to those envars. Some clarification would be greatly appreciated.

Reply:
> When I look through the Open MPI code base, I can see current Open MPI code call `setenv` (or `opal_setenv`) to set similar envars for PSM, PSM2, HCOLL, PVFS etc.
> 
> It is unclear to me whether or the same non-determinism concern apply to those envars. Some clarification would be greatly appreciated.

I'm not familiar with the specific code, but it could create a non-determinism issue. It depends on how it is used. In the case of this thread, the problem is that libfabric looks at `FI_HMEM_CUDA_ENABLE_XFER` during init and then locks its behavior based on that value. Subsequent calls to init are unable to modify that result, nor can they detect that the lock was set. Hence, OMPI (or the app or other library) uses libfabric without actually knowing its state.

If these folks are setting an envar and using it to establish some internal state, then it can be a problem if some other library (or the app itself) changes the envar and therefore modifies the internal state. OMPI then uses the library thinking it is in one state, while it now is in some entirely different state.

Probably should bring it to their attention. Too often, people think that only one library (OMPI) would use a support lib (e.g., libfabric) and don't even realize that other libs and apps also call those functions. This is what leads to the problems described above.

Reply:
> Not saying I don't prefer the hints->caps method, but the non-determinism, to me, isn't a huge factor for this specific flag. For Open MPI code, there is 0 change either way whether the provider actually reads this flag or not. The flag just allows the provider to optimize certain cases, but if it doesn't it doesn't actually change anything Open MPI needs to do.

Well, if there is no impact inside OMPI, then why bother to set the flag in the first place? I was under the impression that the reason for this PR was to ensure that libfabric was set to a mode that the OMPI code desires. If we actually don't care, then there is no reason for us to set anything - and these problems go away.

If you do care about getting that optimization, then determinism is important. Otherwise, you'll have users seeing their apps change their performance without any real understanding as to why/how that happened - e.g., in the case I reported, all they did was add an unrelated lib to their app (and the lib happened to set the flag to something our lib didn't like - and no way to know it happened). Getting a deterministic method for setting libfabric state reduces your support calls and debugging efforts.